### DISCUSSION TITLE
SLOW IN PYTHON TASK

### DISCUSSION THREAD

**Discussion by biohuangjue** [talmolab/sleap#1933] [Help!]
Hi, I'm a sleap user and successfully build my python program for real-time task, but when we change our working environment to another computer, with a RTX3060, our code can only run at 20FPS, while we can run 90FPS on our previous computer. I checked that the Run Inference in GUI can still run at 90 FPS, why this happens. The code I use to test is paste below #

```py
import cv2
import sleap
import time
predictor = sleap.load_model([r"F:\guangyichuan\models\240727_145201.centroid.n=38.zip", r"F:\guangyichuan\models\240727_150142.centered_instance.n=38.zip"], batch_size=16)  
prev_time = time.time()  
width = 1280
height = 1024
cap = cv2.VideoCapture(r"C:\Users\A\Desktop\guangyichuan\videos\gr5a2.7.3-1.mp4")
cap.set(cv2.CAP_PROP_FRAME_WIDTH, width)  
cap.set(cv2.CAP_PROP_FRAME_HEIGHT, height)
while True:
    ret, frame = cap.read()
    current_time = time.time()
    fps = 1.0 / (current_time - prev_time)
    prev_time = current_time
    cv2.putText(frame, f"FPS: {fps:.2f}", (10, 150), cv2.FONT_HERSHEY_SIMPLEX, 1, (255, 255, 255), 2)
    cv2.imshow('Camera Feed', frame)
    key = cv2.waitKey(1)`
```

**Comment by Shriharsh-Deshmukh** [comment]
The issue might be related to how the RTX 3060 is handling the workload. Although the GUI runs inference at 90 FPS, your custom Python code might not be utilizing the GPU efficiently. Ensure that:

1. GPU drivers are up to date.
2. CUDA and cuDNN are properly installed.
3. SLEAP is configured to use the GPU.
4. Batch size is optimized for the new environment.
5. Consider using `cv2.setUseOptimized(True)` to boost OpenCV performance.

If the issue persists, compare the hardware specs and CUDA versions between the two systems.

**Reply by biohuangjue** [reply]
While I tested my GPU functions normally on other task, and find it only can't work in sleap task including inference task and training task. The usage of my GPU is 0%, but the task will take 75% of the memory of my GPU, how can I check whether SLEAP is configured to use the GPU? My team member has checked that the task is set to use GPU. I found a former discussion solves this by seperating CPU and GPU workload, but how to do so?
<img width="583" alt="73c8e9f0c42cb9bede473d1415798af" src="https://github.com/user-attachments/assets/61f7152a-4ddc-4ede-b225-19f1531d24d6">

**Reply by eberrigan** [reply]
Hi @biohuangjue,

How did you install SLEAP?

Best,

Elizabeth

**Reply by biohuangjue** [reply]
I installed SLEAP as recommended on your website. I created a conda environment and installed sleap v1.3.3

**Comment by talmo** [comment]
Hi @biohuangjue,

Are you sure you pasted the right script? That looks like it's just reading frames from a video and displaying it.

Here's something that actually calls the inference

```py
import cv2
import sleap
import time
predictor = sleap.load_model([r"F:\guangyichuan\models\240727_145201.centroid.n=38.zip", r"F:\guangyichuan\models\240727_150142.centered_instance.n=38.zip"], batch_size=16)  
prev_time = time.time()  
width = 1280
height = 1024
cap = cv2.VideoCapture(r"C:\Users\A\Desktop\guangyichuan\videos\gr5a2.7.3-1.mp4")
cap.set(cv2.CAP_PROP_FRAME_WIDTH, width)  
cap.set(cv2.CAP_PROP_FRAME_HEIGHT, height)
while True:
    ret, frame = cap.read()
    frame = frame.reshape(1, height, width, -1)  # make sure image is rank-4
    preds = predictor.infererence_model.predict_on_batch(frame)
    # do something with predictions?
    current_time = time.time()
    fps = 1.0 / (current_time - prev_time)
    prev_time = current_time
    cv2.putText(frame, f"FPS: {fps:.2f}", (10, 150), cv2.FONT_HERSHEY_SIMPLEX, 1, (255, 255, 255), 2)
    cv2.imshow('Camera Feed', frame)
    key = cv2.waitKey(1)`
```

----

That aside, not sure why you're seeing the discrepancy, but have you isolated it to make sure it's the exact same code running on the same data?

As far as usage in the Windows task manager: this is expected. You might need to select "Cuda" or another category in the dropdowns in the GPU tab, but the % usage next to the specific process is probably always inaccurate since it combines usage across multiple categories (e.g., video reading, desktop graphics, etc.).

**Reply by biohuangjue** [reply]
Sorry I pasted the wrong code, the actual code include`      frame_predictions = self.predictor.inference_model.predict_on_batch(np.expand_dims(frame, axis=0))
`, and the discrepancy is between python ans GUI, the model and the video I used is the same. I wonder if it has anything to do with opencv?

**Reply by talmo** [reply]
Hi @biohuangjue,

Yep, a key difference would certainly be the use of opencv, but technically that's what we do under the hood for video reading as well.

Have you tried timing just the `predict_on_batch` call by itself to see how fast that's going separate from the whole loop?

**Reply by biohuangjue** [reply]
Yes, I've tried it. With cv2.imshow, the latency is about 0.026-0.030, without cv2.imshow, the latency is between 0.016-0.020, thus the fps without cv2.imshow is about 50FPS.
